# Práctica · Spark DataFrames · Clientes y pedidos

Completa las celdas indicadas. Trabaja sobre el entorno dockerizado incluido en `spark_jupyter/`.

In [21]:
from iniciar_spark import get_spark
from pyspark.sql import functions as F

spark = get_spark("PracticaClientesPedidos")
print("SparkSession creada")
print("Versión de Spark:", spark.version)


SparkSession creada
Versión de Spark: 4.1.1


In [22]:
from pathlib import Path

data_dir = Path("/opt/spark-apps/datos")
clientes_path = data_dir / "clientes.csv"
pedidos_path = data_dir / "pedidos.csv"

print("Ruta clientes:", clientes_path)
print("Ruta pedidos:", pedidos_path)


Ruta clientes: /opt/spark-apps/datos/clientes.csv
Ruta pedidos: /opt/spark-apps/datos/pedidos.csv


## 1. Lee ambos CSV y muestra su esquema y algunas filas.

In [23]:
# Lectura del dataset de Clientes
df_clientes = (spark.read
    .option("sep", ";")           # Especificamos el punto y coma como separador de columnas
    .option("header", True)       # Indicamos que la primera fila contiene los nombres de las columnas
    .option("inferSchema", True)  # Spark escanea los datos para detectar los tipos (entero, texto, etc.)
    .csv(str(clientes_path)))     # Convertimos el objeto Path a string para que PySpark lo lea sin fallos

# Verificamos que la carga es correcta mostrando el esquema y las primeras filas
df_clientes.printSchema()
df_clientes.show(15)

# Lectura del dataset de Pedidos (aplicamos exactamente la misma configuración)
df_pedidos = (spark.read
    .option("sep", ";")
    .option("header", True)
    .option("inferSchema", True)
    .csv(str(pedidos_path)))

# Verificamos la carga de pedidos
df_pedidos.printSchema()
df_pedidos.show(15)

root
 |-- id_cliente: integer (nullable = true)
 |-- nombre: string (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- segmento: string (nullable = true)

+----------+------+----------+--------+
|id_cliente|nombre|    ciudad|segmento|
+----------+------+----------+--------+
|         1|   Ana|  Sevilla |Estandar|
|         2|  Luis|    Bilbao| Premium|
|         3| Marta| Alicante |Estandar|
|         4| Pablo|   Madrid | Premium|
|         5| Lucia|    Bilbao| Premium|
|         6|Carlos| Alicante |Estandar|
|         7| Elena|    Bilbao|Estandar|
|         8|Javier|    Madrid| Premium|
|         9|  Sara|    Murcia| Premium|
|        10| David|    Bilbao|Estandar|
|        11|Raquel|  Sevilla | Premium|
|        12|Miguel|  Zaragoza|Estandar|
|        13| Irene|    Madrid| Premium|
|        14|Sergio|   Murcia |Estandar|
|        15| Paula|   Granada|Estandar|
+----------+------+----------+--------+
only showing top 15 rows
root
 |-- id_pedido: integer (nullable = true)
 |-

[Stage 5:>                                                          (0 + 1) / 1]

+---------+----------+----------+-----------+--------+---------------+
|id_pedido|id_cliente|     fecha|   producto|cantidad|precio_unitario|
+---------+----------+----------+-----------+--------+---------------+
|     1001|        12|2025-03-05|      Ratón|     6.0|             16|
|     1002|        32|2025-02-02|      Ratón|     3.0|             19|
|     1003|         5|2025-03-07|    Monitor|     2.0|            210|
|     1004|        13|2025-03-18|  Impresora|     4.0|            205|
|     1005|        41|2025-03-10|    Teclado|     5.0|             40|
|     1006|         5|2025-03-05|    Teclado|    NULL|             35|
|     1007|        27|2025-02-13|      Ratón|     1.0|             23|
|     1008|         7|2025-02-04|     Webcam|     6.0|             77|
|     1009|        18|2025-03-02|    Teclado|     1.0|             38|
|     1010|        31|2025-02-14|     Webcam|     1.0|             78|
|     1011|        36|2025-03-15|Auriculares|     2.0|             32|
|     

## 2. Limpia el DataFrame de clientes: trim en ciudad y dropDuplicates.

In [24]:
# --- PERFILADO DE DATOS (DATA PROFILING) ---
# Antes de modificar nada, contamos cuántas filas tenemos en total
# para tener una referencia del tamaño original de nuestro dataset.
total_inicial = df_clientes.count()
print(f"Total de registros iniciales: {total_inicial}")

# Contamos cuántos duplicados exactos hay en origen. 
# La lógica es sencilla: restamos los registros únicos (dropDuplicates) al total de registros.
total_distintos = df_clientes.dropDuplicates().count()
print(f"Número de filas duplicadas antes de limpiar: {total_inicial - total_distintos}")

# Calculamos cuántos valores nulos hay en cada columna para conocer la "salud" de los datos.
# ¿Cómo funciona? isNull() da True/False, cast("int") lo pasa a 1/0, y sum() suma esos unos.
print("Recuento de nulos por columna ANTES de limpiar:")
df_clientes.select(
    *[F.sum(F.col(c).isNull().cast("int")).alias(c) for c in df_clientes.columns]
).show()
# --------------------------------------------------

# 1. Validar tipos de datos: 
# Forzamos que 'id_cliente' sea un número entero. Esto evita errores futuros 
# si inferSchema interpretó algún número con espacios raros como texto.
df_clientes = df_clientes.withColumn("id_cliente", F.col("id_cliente").cast("integer"))

# 2. Gestionar valores nulos: 
# - dropna: Eliminamos la fila completa si falta el 'id_cliente', ya que es nuestra clave principal para el join.
# - fillna: Si falta el nombre, la ciudad o el segmento, no borramos la fila, rellenamos el hueco con "Desconocido".
df_clientes = df_clientes.dropna(subset=["id_cliente"])
df_clientes = df_clientes.fillna("Desconocido", subset=["nombre", "ciudad", "segmento"])

# 3. Eliminar espacios en columnas de texto:
# F.trim limpia los espacios en blanco invisibles al principio y al final de los textos. 
# Así evitamos que Spark considere " Madrid" y "Madrid " como dos ciudades distintas.
df_clientes = (df_clientes
    .withColumn("nombre", F.trim(F.col("nombre")))
    .withColumn("ciudad", F.trim(F.col("ciudad")))
    .withColumn("segmento", F.trim(F.col("segmento")))
)

# 4. Eliminar duplicados:
# Aplicamos la eliminación real de los duplicados que detectamos arriba en el perfilado.
df_clientes = df_clientes.dropDuplicates()

# ==============================================================================
# Verificamos los resultados tras la limpieza
# Hacemos un nuevo count() para saber cuántos registros han sobrevivido.
total_final = df_clientes.count()
print(f"Total de registros DESPUÉS de limpiar: {total_final}")

# Mostramos el total exacto de filas eliminadas (por ser nulos críticos o duplicados)
print(f"Registros eliminados (nulos críticos o duplicados): {total_inicial - total_final}")

print("Datos limpios:")
df_clientes.show(15) # Mostramos 15 filas para tener una mejor visión general del resultado

Total de registros iniciales: 43
Número de filas duplicadas antes de limpiar: 3
Recuento de nulos por columna ANTES de limpiar:
+----------+------+------+--------+
|id_cliente|nombre|ciudad|segmento|
+----------+------+------+--------+
|         0|     0|     0|       0|
+----------+------+------+--------+

Total de registros DESPUÉS de limpiar: 40
Registros eliminados (nulos críticos o duplicados): 3
Datos limpios:
+----------+--------+--------+--------+
|id_cliente|  nombre|  ciudad|segmento|
+----------+--------+--------+--------+
|        18|   Jorge| Sevilla| Premium|
|        40|   Tomas|Valencia|Estandar|
|         5|   Lucia|  Bilbao| Premium|
|         8|  Javier|  Madrid| Premium|
|        17|    Alba|  Madrid|Estandar|
|         2|    Luis|  Bilbao| Premium|
|        33|Patricia|Alicante| Premium|
|        27|   Laura|  Bilbao|Estandar|
|        32|  Andres|Zaragoza| Premium|
|        21|  Noelia|  Malaga| Premium|
|        39|     Eva|Zaragoza|Estandar|
|        12|  Miguel

## 3. Convierte tipos en pedidos y crea la columna `importe`.

In [25]:
# (Incluye limpieza completa: trim, duplicados y tratamiento de nulos)
# ==============================================================================

# --- PERFILADO DE DATOS (DATA PROFILING) ---
# Guardamos el total de filas original para compararlo al final del proceso.
total_inicial_pedidos = df_pedidos.count()
print(f"Total de registros iniciales en pedidos: {total_inicial_pedidos}")

# Calculamos los duplicados exactos restando los registros únicos al total.
total_distintos_pedidos = df_pedidos.dropDuplicates().count()
print(f"Número de filas duplicadas antes de limpiar: {total_inicial_pedidos - total_distintos_pedidos}")

# Radiografía de nulos: Vemos cuántos valores faltan en cada columna.
# ¿Cómo funciona? isNull() da True/False, cast("int") lo pasa a 1/0, y sum() suma esos unos.
print("Recuento de nulos por columna ANTES de limpiar:")
df_pedidos.select(
    *[F.sum(F.col(c).isNull().cast("int")).alias(c) for c in df_pedidos.columns]
).show()
# --------------------------------------------------

# 1. Validar y convertir tipos de datos:
# Aunque usamos 'inferSchema=True' al leer el CSV, siempre es mejor forzar los tipos manualmente
# para evitar que un carácter extraño convierta toda una columna numérica en texto.
df_pedidos = (df_pedidos
    .withColumn("id_pedido", F.col("id_pedido").cast("integer"))         # Forzamos a número entero
    .withColumn("id_cliente", F.col("id_cliente").cast("integer"))       # Clave para el join futuro
    .withColumn("fecha", F.to_date(F.col("fecha")))                      # Convierte texto a formato Date nativo de Spark
    .withColumn("cantidad", F.col("cantidad").cast("double"))            # Double permite decimales
    .withColumn("precio_unitario", F.col("precio_unitario").cast("double")) 
)

# 2. Gestionar valores nulos:
# - dropna: Si un pedido no tiene 'id_pedido' o no sabemos de qué cliente es ('id_cliente'),
#   ese registro es "basura" porque no podremos cruzarlo. Lo eliminamos por completo.
df_pedidos = df_pedidos.dropna(subset=["id_pedido", "id_cliente"])

# - fillna (textos): Si nos falta el nombre del producto, no borramos la venta, simplemente 
#   le ponemos la etiqueta "Desconocido" para no perder el importe económico de esa fila.
df_pedidos = df_pedidos.fillna("Desconocido", subset=["producto"])

# - fillna (números): Si falta la cantidad o el precio, ponemos un 0.0. 
#   ESTO ES CRÍTICO: Si intentamos multiplicar un número por un valor nulo (Null), Spark 
#   devolverá Null, y perderemos el cálculo. Con un 0.0, evitamos que el código falle.
df_pedidos = df_pedidos.fillna(0.0, subset=["cantidad", "precio_unitario"])

# 3. Eliminar espacios en columnas de texto (Trim):
# F.trim limpia espacios invisibles a los lados del texto.
# Evita que el sistema de análisis se confunda pensando que "Monitor" y " Monitor " son dos productos distintos.
df_pedidos = df_pedidos.withColumn("producto", F.trim(F.col("producto")))

# 4. Eliminar duplicados:
# Limpiamos las filas que sean 100% idénticas al resto.
df_pedidos = df_pedidos.dropDuplicates()

# 5. CREAR LA COLUMNA IMPORTE:
# Spark hace esta operación matemática de forma distribuida. Multiplica fila por fila
# la columna 'cantidad' por 'precio_unitario' y guarda el resultado en una nueva columna llamada 'importe'.
df_pedidos = df_pedidos.withColumn("importe", F.col("cantidad") * F.col("precio_unitario"))

# ==============================================================================
# Verificamos los resultados tras la limpieza y transformación
total_final_pedidos = df_pedidos.count()
print(f"\nTotal de registros DESPUÉS de limpiar: {total_final_pedidos}")
print(f"Registros eliminados (nulos críticos o duplicados): {total_inicial_pedidos - total_final_pedidos}")

# Mostramos el esquema para verificar que los tipos (integer, date, double) están correctos
# y que la columna 'importe' se ha creado con éxito.
print("\nEsquema final validado:")
df_pedidos.printSchema()

print("Datos limpios y transformados:")
df_pedidos.show(15)

Total de registros iniciales en pedidos: 120
Número de filas duplicadas antes de limpiar: 0
Recuento de nulos por columna ANTES de limpiar:
+---------+----------+-----+--------+--------+---------------+
|id_pedido|id_cliente|fecha|producto|cantidad|precio_unitario|
+---------+----------+-----+--------+--------+---------------+
|        0|         0|    0|       2|       3|              0|
+---------+----------+-----+--------+--------+---------------+


Total de registros DESPUÉS de limpiar: 120
Registros eliminados (nulos críticos o duplicados): 0

Esquema final validado:
root
 |-- id_pedido: integer (nullable = true)
 |-- id_cliente: integer (nullable = true)
 |-- fecha: date (nullable = true)
 |-- producto: string (nullable = false)
 |-- cantidad: double (nullable = false)
 |-- precio_unitario: double (nullable = false)
 |-- importe: double (nullable = false)

Datos limpios y transformados:
+---------+----------+----------+-----------+--------+---------------+-------+
|id_pedido|id_c

## 4. Haz un join entre clientes y pedidos.

In [29]:
# --- ANÁLISIS DE INTEGRIDAD (PRE-JOIN) ---
# Igual que hicimos en la limpieza, antes de cruzar tablas es vital saber 
# cuántos registros tenemos en origen para medir el impacto del cruce.
count_clientes = df_clientes.count()
count_pedidos = df_pedidos.count()
print(f"Registros antes del cruce: {count_clientes} Clientes y {count_pedidos} Pedidos.")

# 1. REALIZAR EL CRUCE PRINCIPAL (INNER JOIN):
# Un 'inner join' es como buscar coincidencias exactas. Spark mira la columna 
# 'id_cliente' en ambas tablas. Si el ID existe en las dos, une los datos en una sola fila.
# OJO: Si un ID está en clientes pero no en pedidos (o viceversa), esa fila SE PIERDE.
# El resultado se guarda en un nuevo DataFrame llamado 'df_ventas'.
df_ventas = df_clientes.join(df_pedidos, on="id_cliente", how="inner")

# 2. ANÁLISIS DE REGISTROS PERDIDOS (Requisito del proyecto):
# Como sabemos que el 'inner join' elimina los que no coinciden perfectamente, 
# vamos a usar una técnica analítica llamada 'Anti Join' (left_anti).
# Un left_anti join devuelve SOLO las filas de la tabla de la izquierda que NO 
# encuentran a su "pareja" en la tabla de la derecha.

# A. Encontrar pedidos "huérfanos":
# Ponemos df_pedidos a la izquierda. Si un pedido tiene un 'id_cliente' que NO 
# existe en la tabla df_clientes, el Anti Join lo captura. Esto indica un error de datos.
pedidos_sin_cliente = df_pedidos.join(df_clientes, on="id_cliente", how="left_anti")

# B. Encontrar clientes "inactivos" (que no han comprado):
# Ponemos df_clientes a la izquierda. Si un cliente no tiene ningún pedido a su 
# nombre en df_pedidos, el Anti Join lo captura. Es un cliente registrado sin ventas.
clientes_sin_pedido = df_clientes.join(df_pedidos, on="id_cliente", how="left_anti")

# ==============================================================================
# RESULTADOS Y CONCLUSIONES DEL CRUCE
# Imprimimos un pequeño informe para ver exactamente qué ha pasado en el cruce.
# ==============================================================================
print("-" * 50)
print(f"RESULTADO DEL INNER JOIN: {df_ventas.count()} ventas consolidadas.")
print("REGISTROS DESCARTADOS DURANTE EL CRUCE:")
print(f" - Pedidos perdidos (El cliente no existe en la base de datos): {pedidos_sin_cliente.count()}")
print(f" - Clientes descartados (No tienen ninguna compra asociada): {clientes_sin_pedido.count()}")
print("-" * 50)

# Mostramos los pedidos que han quedado huérfanos para poder investigar el problema
if pedidos_sin_cliente.count() > 0:
    print("\nEjemplo de pedidos huérfanos (se requiere investigar estos IDs):")
    pedidos_sin_cliente.select("id_pedido", "id_cliente", "producto").show(15)

# Mostramos cómo queda la tabla final unificada lista para el análisis de negocio
print("\nVista previa del DataFrame unificado (df_ventas) listo para analizar:")
# Seleccionamos un orden de columnas lógico para que la tabla sea fácil de leer en pantalla
df_ventas.select("id_cliente", "nombre", "ciudad", "id_pedido", "producto", "importe").show(15)

Registros antes del cruce: 40 Clientes y 120 Pedidos.
--------------------------------------------------
RESULTADO DEL INNER JOIN: 110 ventas consolidadas.
REGISTROS DESCARTADOS DURANTE EL CRUCE:
 - Pedidos perdidos (El cliente no existe en la base de datos): 10
 - Clientes descartados (No tienen ninguna compra asociada): 3
--------------------------------------------------

Ejemplo de pedidos huérfanos (se requiere investigar estos IDs):
+---------+----------+-----------+
|id_pedido|id_cliente|   producto|
+---------+----------+-----------+
|     1074|        42|  Impresora|
|     1012|        48|Desconocido|
|     1107|        47|    Monitor|
|     1114|        46|      Ratón|
|     1109|        46|Auriculares|
+---------+----------+-----------+
only showing top 5 rows

Vista previa del DataFrame unificado (df_ventas) listo para analizar:
+----------+-------+--------+---------+---------+-------+
|id_cliente| nombre|  ciudad|id_pedido| producto|importe|
+----------+-------+--------+--

## 5. Filtra ventas Premium con importe >= 100.

In [30]:
# --- ANÁLISIS DE SEGMENTACIÓN (FILTERING) ---
# En esta fase, reducimos todos los datos a un subconjunto específico.
# El objetivo es aislar las ventas de clientes 'Premium' que han gastado 100 o más.

# Definimos el filtrado con dos condiciones simultáneas:
# 1. 'segmento' debe ser exactamente igual a "Premium".
# 2. 'importe' (columna creada en la celda 3) debe ser mayor o igual a 100.
# Nota: Usamos el símbolo '&' para indicar que AMBAS condiciones deben cumplirse.
df_premium_top = df_ventas.filter(
    (F.col("segmento") == "Premium") & 
    (F.col("importe") >= 100)
)

# ==============================================================================
# VERIFICACIÓN DE RESULTADOS
# ==============================================================================

# Contamos cuántos registros cumplen con este criterio de negocio
count_premium = df_premium_top.count()
print(f"Número de ventas Premium de alto valor (>= 100): {count_premium}")

# Mostramos los resultados obtenidos
if count_premium > 0:
    print("\nDetalle de las ventas filtradas:")
    # Seleccionamos las columnas más relevantes para validar el filtro visualmente
    df_premium_top.select("nombre", "ciudad", "segmento", "producto", "importe").show()
else:
    print("\nNo se han encontrado registros que cumplan ambas condiciones simultáneamente.")

# Pequeña auditoría: Verificamos si hay algún nulo en las columnas filtradas 
# que pudiera haber afectado al resultado.
print("\nVerificación de integridad en el filtro:")
df_premium_top.select(F.count(F.when(F.col("segmento").isNull(), 1)).alias("Nulos_Segmento")).show()

Número de ventas Premium de alto valor (>= 100): 34

Detalle de las ventas filtradas:
+--------+--------+--------+---------+-------+
|  nombre|  ciudad|segmento| producto|importe|
+--------+--------+--------+---------+-------+
|   Oscar|Alicante| Premium| Portátil| 3705.0|
|   Jorge| Sevilla| Premium|   Webcam|  252.0|
|   Pablo|  Madrid| Premium|   Webcam|  159.0|
|   Oscar|Alicante| Premium|Impresora| 1146.0|
|   Lucia|  Bilbao| Premium|Disco SSD|  105.0|
|Patricia|Alicante| Premium|Disco SSD|  246.0|
|   Diego| Granada| Premium|Impresora|  748.0|
|   Oscar|Alicante| Premium|   Webcam|  215.0|
|   Pablo|  Madrid| Premium|  Monitor|  870.0|
|   Diego| Granada| Premium|   Webcam|  156.0|
|    Sara|  Murcia| Premium| Portátil|  983.0|
|   Ruben| Granada| Premium| Portátil| 6186.0|
|    Luis|  Bilbao| Premium|Impresora| 1015.0|
|   Oscar|Alicante| Premium|Impresora|  140.0|
|  Adrian|Valencia| Premium|   Webcam|  486.0|
|Patricia|Alicante| Premium|Disco SSD|  468.0|
|    Luis|  Bilbao| P

## 6. Clasifica pedidos con `when` en Alto / Medio / Bajo.

In [32]:
# --- TRANSFORMACIÓN CONDICIONAL ---
# Vamos a crear una nueva columna llamada 'categoria_pedido' basada en el 'importe'.
# Funciona leyendo de arriba a abajo: si cumple la primera condición, asigna el valor y pasa a la siguiente fila.
# Si no la cumple, evalúa el siguiente '.when()'. Si no cumple ninguna, cae en el '.otherwise()'.

# Umbrales definidos:
# - Alto: importe >= 100
# - Medio: importe entre 50 y 99.99
# - Bajo: importe menor a 50

df_ventas = df_ventas.withColumn(
    "categoria_pedido",
    F.when(F.col("importe") >= 100, "Alto")
     .when(F.col("importe") >= 50, "Medio")
     .otherwise("Bajo") # Si no es Alto ni Medio, por descarte es Bajo
)

# ==============================================================================
# VERIFICACIÓN DE RESULTADOS Y PERFILADO
# ==============================================================================

print("Clasificación aplicada con éxito. Vista previa de la categorización:")
# Mostramos las columnas clave para comprobar que la lógica matemática ha funcionado bien
df_ventas.select("id_pedido", "producto", "importe", "categoria_pedido").show(15)

# --- AUDITORÍA DE DISTRIBUCIÓN ---
# No solo miramos las primeras filas, queremos ver 
# cómo se distribuyen TODOS los datos tras aplicar nuestra nueva regla de negocio.
print("\nDistribución total de pedidos por categoría:")
df_ventas.groupBy("categoria_pedido").count().orderBy(F.desc("count")).show()

Clasificación aplicada con éxito. Vista previa de la categorización:
+---------+---------+-------+----------------+
|id_pedido| producto|importe|categoria_pedido|
+---------+---------+-------+----------------+
|     1013| Portátil| 3705.0|            Alto|
|     1024|   Webcam|  252.0|            Alto|
|     1049|   Webcam|   88.0|           Medio|
|     1031| Portátil| 1886.0|            Alto|
|     1110|Impresora|  798.0|            Alto|
|     1015|  Teclado|  145.0|            Alto|
|     1018|   Webcam|  159.0|            Alto|
|     1100|  Teclado|   90.0|           Medio|
|     1061| Portátil| 4304.0|            Alto|
|     1113|Impresora|  324.0|            Alto|
|     1095|    Ratón|   42.0|            Bajo|
|     1029|Impresora|  406.0|            Alto|
|     1118|Disco SSD|  133.0|            Alto|
|     1079|Impresora| 1146.0|            Alto|
|     1106| Portátil| 1708.0|            Alto|
+---------+---------+-------+----------------+
only showing top 15 rows

Distribución

## 7. Calcula agregaciones por ciudad y segmento.

In [33]:
# --- ANÁLISIS AGREGADO (GROUP BY + AGG) ---
# Objetivo: Resumir el rendimiento del negocio para entender qué ciudades y 
# qué tipos de clientes (segmentos) son los más rentables.

# 1. Agrupamos por nuestras dos dimensiones clave: ciudad y segmento.
df_metricas = (df_ventas
    .groupBy("ciudad", "segmento")
    
    # 2. Aplicamos las métricas matemáticas (Agregaciones)
    .agg(
        # Métrica 1: Número de pedidos (Cuenta cuántas filas/pedidos hay en el grupo)
        # Usamos alias() para que la columna no se llame "count(id_pedido)"
        F.count("id_pedido").alias("num_pedidos"),
        
        # Métrica 2: Ingresos totales (Suma el importe de todos los pedidos del grupo)
        # Envolvemos la suma en F.round(..., 2) para dejar solo dos decimales
        F.round(F.sum("importe"), 2).alias("ingresos_totales"),
        
        # Métrica 3: Medias (Calcula el ticket medio o gasto promedio por pedido)
        F.round(F.avg("importe"), 2).alias("ticket_medio_pedido")
    )
    
    # 3. Ordenamos de mayor a menor beneficio para que la tabla sea fácil de interpretar
    .orderBy(F.desc("ingresos_totales"))
)

# ==============================================================================
# RESULTADOS DEL ANÁLISIS ESTRATÉGICO
# ==============================================================================

print("Métricas de negocio consolidadas por Ciudad y Segmento:")
print("(Ordenado por volumen de ingresos totales)\n")

# Mostramos el resultado final.
df_metricas.show()

# Breve auditoría final: ¿Cuántos grupos o perfiles distintos de clientes tenemos?
print(f"Se han identificado {df_metricas.count()} combinaciones únicas de Ciudad-Segmento con ventas.")

Métricas de negocio consolidadas por Ciudad y Segmento:
(Ordenado por volumen de ingresos totales)

+----------+--------+-----------+----------------+-------------------+
|    ciudad|segmento|num_pedidos|ingresos_totales|ticket_medio_pedido|
+----------+--------+-----------+----------------+-------------------+
|    Bilbao|Estandar|         19|         11685.0|              615.0|
|   Granada| Premium|          7|          9161.0|            1308.71|
|  Alicante|Estandar|         10|          9113.0|              911.3|
|  Alicante| Premium|          7|          7472.0|            1067.43|
|    Bilbao| Premium|          8|          5674.0|             709.25|
|   Sevilla|Estandar|          6|          3091.0|             515.17|
|    Madrid|Estandar|          5|          2719.0|              543.8|
|  Zaragoza|Estandar|          6|          2710.0|             451.67|
|    Madrid| Premium|          5|          2347.0|              469.4|
|    Murcia|Estandar|          8|          2058.

## 8. Crea una vista temporal y haz una consulta SQL.

In [35]:
# 1. CREACIÓN DE LA VISTA TEMPORAL
# Para poder usar SQL, primero debemos registrar nuestro DataFrame como una 'tabla' virtual.
# 'createOrReplaceTempView' le dice a Spark: "A partir de ahora, cuando en un SQL escriba 
# 'ventas_reporte', me refiero a los datos que hay dentro de df_ventas".
# Usamos 'OrReplace' por seguridad: si ejecutamos la celda dos veces, simplemente sobrescribe la vista.
df_ventas.createOrReplaceTempView("ventas_reporte")

# 2. EJECUCIÓN DE CONSULTA SQL SIGNIFICATIVA
# Vamos a realizar un análisis que sería complejo de leer en Python pero que en SQL es muy natural:
# "Queremos el Top 5 de productos que más dinero han generado, junto con su precio medio 
# y cuántas unidades totales se han vendido."

query_sql = """
    SELECT 
        producto,
        SUM(cantidad) AS unidades_totales,
        ROUND(AVG(precio_unitario), 2) AS precio_medio,
        ROUND(SUM(importe), 2) AS recaudacion_total
    FROM 
        ventas_reporte
    GROUP BY 
        producto
    HAVING 
        recaudacion_total > 0
    ORDER BY 
        recaudacion_total DESC
    LIMIT 5
"""

# Ejecutamos la consulta usando spark.sql()
# El resultado de un spark.sql() SIEMPRE es otro DataFrame de Spark.
df_sql_resultados = spark.sql(query_sql)

# ==============================================================================
# VERIFICACIÓN DE RESULTADOS SQL
# ==============================================================================

print("Análisis de Productos Top (Generado mediante Spark SQL):")
print("-" * 50)

# Mostramos los resultados obtenidos por la consulta SQL
df_sql_resultados.show()

Análisis de Productos Top (Generado mediante Spark SQL):
--------------------------------------------------
+---------+----------------+------------+-----------------+
| producto|unidades_totales|precio_medio|recaudacion_total|
+---------+----------------+------------+-----------------+
| Portátil|            33.0|      971.42|          31829.0|
|Impresora|            83.0|      165.32|          13315.0|
|Disco SSD|            56.0|       94.53|           5349.0|
|   Webcam|            74.0|       54.94|           4057.0|
|  Monitor|            19.0|      210.33|           3853.0|
+---------+----------------+------------+-----------------+



## 9. Usa `sample()` y `randomSplit()`.

In [37]:
# --- PREPARACIÓN PARA MACHINE LEARNING Y TESTING ---
# En bases de datos masivas, necesitamos extraer subconjuntos aleatorios para 
# entrenar algoritmos predictivos o simplemente para probar código rápidamente.

# 1. USO DE SAMPLE()
# Extraemos una muestra aleatoria de aproximadamente el 20% (0.2) de las ventas consolidadas.
# - withReplacement=False: Una misma fila no puede ser extraída dos veces en la muestra.
# - seed=42: Fijamos la "semilla". Esto asegura que, aunque sea aleatorio, cada vez 
#   que se ejecute la celda, saldrán exactamente los mismos registros.
df_muestra = df_ventas.sample(withReplacement=False, fraction=0.2, seed=42)

# 2. USO DE RANDOMSPLIT()
# Dividimos el DataFrame completo en dos pedazos distintos: 
# - Un 80% (0.8) que usaríamos para "Entrenar" (Train) un algoritmo.
# - Un 20% (0.2) que usaríamos para "Testear" (Test) el algoritmo.
# Spark se encarga de repartir las filas aleatoriamente respetando estos porcentajes.
df_train, df_test = df_ventas.randomSplit([0.8, 0.2], seed=42)

# ==============================================================================
# VERIFICACIÓN DE RESULTADOS Y AUDITORÍA
# ==============================================================================

total_ventas = df_ventas.count()

print("--- AUDITORÍA DE DIVISIONES ALEATORIAS ---")
print(f"Total de registros originales (Universo): {total_ventas}\n")

print("1. RESULTADOS DE SAMPLE():")
print(f" - Registros en la muestra (aprox 20%): {df_muestra.count()}")

print("\n2. RESULTADOS DE RANDOMSPLIT():")
print(f" - Registros asignados a Entrenamiento (aprox 80%): {df_train.count()}")
print(f" - Registros asignados a Testing (aprox 20%): {df_test.count()}")

# Verificación de integridad: La suma de train + test debe ser IGUAL al total original
total_split = df_train.count() + df_test.count()
print(f" - Suma de Train + Test: {total_split} (¿Coincide con el total?: {total_ventas == total_split})")
print("-" * 50)

# ==============================================================================
# VISUALIZACIÓN DE LOS DATASETS CREADOS
# ==============================================================================

if df_muestra.count() > 0:
    print("\nVista previa de la MUESTRA extraída (sample):")
    df_muestra.select("id_pedido", "id_cliente", "producto", "importe").show(3)

if df_train.count() > 0:
    print("\nVista previa de datos de ENTRENAMIENTO (df_train):")
    df_train.select("id_pedido", "id_cliente", "producto", "importe").show(3)

if df_test.count() > 0:
    print("\nVista previa de datos de TESTING (df_test):")
    df_test.select("id_pedido", "id_cliente", "producto", "importe").show(3)

--- AUDITORÍA DE DIVISIONES ALEATORIAS ---
Total de registros originales (Universo): 110

1. RESULTADOS DE SAMPLE():
 - Registros en la muestra (aprox 20%): 21

2. RESULTADOS DE RANDOMSPLIT():
 - Registros asignados a Entrenamiento (aprox 80%): 92
 - Registros asignados a Testing (aprox 20%): 18
 - Suma de Train + Test: 110 (¿Coincide con el total?: True)
--------------------------------------------------

Vista previa de la MUESTRA extraída (sample):
+---------+----------+---------+-------+
|id_pedido|id_cliente| producto|importe|
+---------+----------+---------+-------+
|     1100|        23|  Teclado|   90.0|
|     1045|         5|Disco SSD|  105.0|
|     1111|        33|Disco SSD|  246.0|
+---------+----------+---------+-------+
only showing top 3 rows

Vista previa de datos de ENTRENAMIENTO (df_train):
+---------+----------+---------+-------+
|id_pedido|id_cliente| producto|importe|
+---------+----------+---------+-------+
|     1029|         1|Impresora|  406.0|
|     1037|      

## 10. Guarda el resultado en Parquet en `/opt/spark-apps/salida/resultado_parquet` y léelo de nuevo.

In [38]:
# Definimos la ruta de salida especificada en el enunciado.
ruta_parquet = "/opt/spark-apps/salida/resultado_parquet"

# 1. PERSISTENCIA: GUARDAR EN FORMATO PARQUET
# Parquet es el estándar de oro en Big Data. A diferencia del CSV (que es texto plano),
# Parquet es un formato binario y 'columnar'.
# - Comprime los datos drásticamente (ahorra espacio).
# - Guarda los tipos de datos (no hay que volver a usar 'cast' al leer).
# - Permite leer solo las columnas necesarias, lo que es muchísimo más rápido.

print(f"Guardando los resultados finales en: {ruta_parquet}...")

# Usamos .write con el modo 'overwrite' para que, si se ejecuta la celda varias veces, 
# Spark sobrescriba la carpeta sin dar un error porque la carpeta ya existe.
df_ventas.write.mode("overwrite").parquet(ruta_parquet)

print("Archivo guardado con éxito en formato binario Parquet")

# 2. PERSISTENCIA: LEER EL ARCHIVO DE NUEVO
# Para verificar que el guardado fue correcto, 
# cargamos los datos desde el disco a un nuevo DataFrame.
print("\nLeyendo los datos de nuevo desde el archivo Parquet para verificar...")

# Al leer Parquet, Spark recupera automáticamente el esquema y los nombres de columnas.
df_final_leido = spark.read.parquet(ruta_parquet)

# ==============================================================================
# VERIFICACIÓN FINAL
# ==============================================================================

# Comprobamos que el esquema se ha mantenido perfectamente (id es integer, fecha es date, etc.)
print("Esquema recuperado (Persistencia de tipos verificada):")
df_final_leido.printSchema()

# Mostramos los datos finales que han pasado por todo el pipeline: 
# Limpieza -> Join -> Transformación -> Persistencia.
print("Muestra de los datos recuperados del disco:")
df_final_leido.show(5)

Guardando los resultados finales en: /opt/spark-apps/salida/resultado_parquet...


Archivo guardado con éxito en formato binario Parquet

Leyendo los datos de nuevo desde el archivo Parquet para verificar...
Esquema recuperado (Persistencia de tipos verificada):
root
 |-- id_cliente: integer (nullable = true)
 |-- nombre: string (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- segmento: string (nullable = true)
 |-- id_pedido: integer (nullable = true)
 |-- fecha: date (nullable = true)
 |-- producto: string (nullable = true)
 |-- cantidad: double (nullable = true)
 |-- precio_unitario: double (nullable = true)
 |-- importe: double (nullable = true)
 |-- categoria_pedido: string (nullable = true)

Muestra de los datos recuperados del disco:
+----------+------+--------+--------+---------+----------+---------+--------+---------------+-------+----------------+
|id_cliente|nombre|  ciudad|segmento|id_pedido|     fecha| producto|cantidad|precio_unitario|importe|categoria_pedido|
+----------+------+--------+--------+---------+----------+---------+--------+-----

## 11. Respuestas de Análisis Teórico

* **¿Qué ventaja tiene usar join?**
  La principal ventaja es la **integración y enriquecimiento de datos**. Al cruzar tablas, transformamos IDs numéricos sin contexto en información de negocio procesable. Por ejemplo, en nuestro análisis (Celda 4), gracias al join pudimos descubrir que la venta del pedido 1013 (un Portátil de 3705€) fue realizada por Oscar, un cliente de segmento Premium de Alicante. Sin el join, esta segmentación estratégica habría sido imposible.

* **¿Qué diferencia hay entre sample y randomSplit?**
  * **`sample()` (Muestreo):** Extrae un único subconjunto aleatorio (una fracción) de la tabla original para pruebas rápidas. En nuestro caso (Celda 9), nos devolvió 21 filas aleatorias. No está diseñado para conservar o dividir el 100% de la información.
  * **`randomSplit()` (Partición):** Divide el *100% de la tabla original* en dos o más DataFrames mutuamente excluyentes, sin duplicar ni perder un solo registro. En nuestro proyecto, partió exactamente nuestras 110 ventas en 92 registros para Entrenamiento (Train) y 18 para Prueba (Test), un paso crítico y obligatorio para entrenar modelos de Machine Learning sin riesgo de sobreajuste.

* **¿Qué pedido se pierde en el inner join y por qué?**
  En nuestro `inner join` se perdieron exactamente **10 pedidos y 3 clientes** debido a la falta de integridad referencial. 
  1. **Pedidos huérfanos (10):** Se perdieron pedidos (como los IDs 1012, 1074 o 1107) porque estaban asociados a clientes inexistentes en nuestra base de datos maestra (IDs 42, 46, 47, 48...). Al no encontrar una coincidencia en la tabla de clientes, Spark los descarta.
  2. **Clientes inactivos (3):** Se perdieron 3 clientes que, aunque estaban correctamente registrados en la base de datos (del 1 al 40), no tenían ningún pedido asociado a su ID en la tabla de ventas.

## 12. Cierra la sesión de Spark.

In [41]:
spark.stop()